erc721-tests/contracts/Erc-721.sol

```diff
@@ -0,0 +1,147 @@
+    }
+
+    function mintToken(address recipient, string memory name_) onlyOwner public {
+        require(owner()!=recipient, "Recipient cannot be the owner of the contract");
+        _safeMint(recipient, tokenId);
+        ownershipRecord[recipient].push(TokenMetaData(name_, tokenId, 0, 0, tokenURI(tokenId), 0, 0));
+        tokenId = tokenId + 1;
+    }
+
+    function transfer(
+        address from_,
+        address to_,
+        uint256 tokenId_
+    ) external virtual isNotLocked(tokenId) {
+        _transfer(from_, to_, tokenId_);
+    }
+
+    function burn(uint256 tokenIdToBurn, int256 tokenIdToTransferExp) external virtual isNotLocked(tokenId) {
+        require(_ownerOf(tokenIdToBurn) != address(0), "ERC721: invalid token ID");
+        require(tokenIdToTransferExp > 0, "ERC721: invalid reciver token ID");
+
+        uint senderexp = (getTockenMetadata(tokenIdToBurn).experience / 100) * 80;
+        getTockenMetadata(uint(tokenIdToTransferExp)).experience += senderexp;
+
+        _burn(tokenIdToBurn);
+    }
+
+    function getTockenMetadata(uint256 tokenId_) public view returns(TokenMetaData memory) {
+        address tokenAddress = _ownerOf(tokenId_);
+        TokenMetaData[] memory dataList = ownershipRecord[tokenAddress];
+        for (uint256 i = 0; i < dataList.length; i++) {
+            if (dataList[i].tokenId == tokenId_) {
+                return dataList[i];
+            }
+        }
+    }
+
+    function lockToken(uint256 tokenId_, uint256 daysToLock) public isNotLocked(tokenId_) {
+        TokenMetaData memory tokenMetaData_ = getTockenMetadata(tokenId_);
+
+        lockRecord[tokenId_] = true;
+        tokenMetaData_ = TokenMetaData(
+            tokenMetaData_.name, tokenId_, block.timestamp, daysToLock, 
+            tokenURI(tokenId_), tokenMetaData_.experience, tokenMetaData_.rank
+        );
+    }
+
+    function unLockToken(uint256 tokenId_) public {
+        TokenMetaData memory tokenMetaData_ = getTockenMetadata(tokenId_);
+        uint256 lockedTimeStamp_ = tokenMetaData_.lockedTimeStamp;
+        uint256 daysToLock_ = tokenMetaData_.daysToLock;
+
+        uint diff = (block.timestamp - lockedTimeStamp_) / 60 / 60 / 24; 
+        bool overLockDeadline = diff >= daysToLock_;
+
+        require(overLockDeadline, "The token lock time is pending ...");
+
+        addExp(tokenId_, diff * 100);
+        lockRecord[tokenId_] = false;
+        tokenMetaData_ = TokenMetaData(
+            tokenMetaData_.name, tokenId_, tokenMetaData_.lockedTimeStamp, 0, 
+            tokenURI(tokenId_), tokenMetaData_.experience, tokenMetaData_.rank
+        );
+    }
+
+    function claimExp(uint256 tokenId_) public {
+        TokenMetaData memory tokenMetaData_ = getTockenMetadata(tokenId_);
+        uint diff = (block.timestamp - tokenMetaData_.lockedTimeStamp) / 60 / 60 / 24; 
+
+        addExp(tokenId_, diff * 100);
+        tokenMetaData_ = TokenMetaData(
```

Comment:
Again (Assigning value to a variable with 'memory' specifier will not persist. Research 'memory' vs 'storage'.)